presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchClient.java

```diff
@@ -0,0 +1,391 @@
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.spi.type.DoubleType.DOUBLE;
+import static com.facebook.presto.spi.type.IntegerType.INTEGER;
+import static com.facebook.presto.spi.type.VarbinaryType.VARBINARY;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.cache.CacheLoader.asyncReloading;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.util.Map.Entry;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+import static java.util.concurrent.TimeUnit.MINUTES;
+
+public class ElasticsearchClient
+{
+    private static final Logger LOG = Logger.get(ElasticsearchClient.class);
+
+    private final ExecutorService executor = newCachedThreadPool(daemonThreadsNamed("elasticsearch-metadata-%s"));
+    private final ObjectMapper objecMapper = new ObjectMapperProvider().get();
+    private final ElasticsearchTableDescriptionProvider tableDescriptions;
+    private final Map<String, TransportClient> clients = new HashMap<>();
+    private final LoadingCache<ElasticsearchTableDescription, List<ColumnMetadata>> tableColumnsCache;
+    private final Duration timeout;
+
+    @Inject
+    public ElasticsearchClient(ElasticsearchTableDescriptionProvider descriptions, Duration requestTimeout)
+            throws IOException
+    {
+        tableDescriptions = requireNonNull(descriptions, "description is null");
+        timeout = requireNonNull(requestTimeout, "requestTimeout is null");
+
+        for (ElasticsearchTableDescription tableDescription : tableDescriptions.getAllTableDescriptions()) {
+            if (!clients.containsKey(tableDescription.getClusterName())) {
+                Settings settings = Settings.builder().put("cluster.name", tableDescription.getClusterName()).build();
+                TransportAddress address = new TransportAddress(InetAddress.getByName(tableDescription.getHost()), tableDescription.getPort());
+                TransportClient client = new PreBuiltTransportClient(settings).addTransportAddress(address);
+                clients.put(tableDescription.getClusterName(), client);
+            }
+        }
+        this.tableColumnsCache = CacheBuilder.newBuilder()
+                .expireAfterWrite(30, MINUTES)
+                .refreshAfterWrite(15, MINUTES)
+                .maximumSize(500)
+                .build(asyncReloading(CacheLoader.from(this::loadColumns), executor));
+    }
+
+    @PreDestroy
+    public void tearDown()
+    {
+        executor.shutdownNow();
+        try (Closer closer = Closer.create()) {
+            for (Entry<String, TransportClient> entry : clients.entrySet()) {
```

Comment:
We can use closer to close the executor:

```
closer.register(() -> executor.shutdown());
```

Reply:
get it, will fix